### PR TITLE
Psionic Power Rolling Rework

### PR DESCRIPTION
**Pull request by VMSolidus**
# Description

This PR implements a long overdue rework of the underlying system for actually giving Psions their powers, by making it a semi-deterministic system that favors people having a small handful of powers, unless said person has inherent modifiers that let them generate them more consistently. A common complaint I received from the new Psionic System is that stations tended to become "Space Hogwarts" given enough time, since people could very easily game the system to consistently get new powers every 20 minutes, until a majority of the station was throwing fireballs at each other.

Now, obtaining new powers gets more difficult the more powers you have, exponentially so. Powers can also declare in their prototype how many "Slots" they occupy, meaning that certain powers can make this even more difficult, or not at all, as desired.

# Changelog

:cl: VMSolidus & Rane
- add: Psionic Power "Rerolls" have been reworked. Psions now have a stat called Potentia, which is increased by sources of Psionic Rerolls such as Space Drugs, Loto Oil, and Random Events. Instead of being an all-or-nothing "D100 with modifiers" roll, whatever you roll is added to your Potentia stat, and when Potentia reaches a certain threshold, it is automatically "Spent" to buy a random new power. Your first power costs 100 Potentia, your 2nd power costs 200, 3rd costs 400, and so on. The more powers you have, the more difficult it is to obtain new ones.
- tweak: Some powers, such as Telepathy and Xenoglossy, do not count as a power for the purpose of Potentia Cost

### CODE REVIEW

File: Content.Server/Abilities/Psionics/PsionicAbilitiesSystem.cs [hunk 1]
```diff
@@ -297,6 +300,24 @@ private void SendFeedbackMessage(EntityUid uid, PsionicPowerPrototype proto, boo
...
+            psionic.PowerSlotsTaken = slotsUsed;
+        }
+
+        private void UpdatePsionicDanger(EntityUid uid, PsionicComponent psionic)
+        {
+            if (psionic.PowerSlotsTaken <= psionic.PowerSlots)
+                return;
+
+            EnsureComp<GlimmerSourceComponent>(uid, out var glimmerSource);
+            glimmerSource.SecondsPerGlimmer = 10 / (psionic.PowerSlotsTaken - psionic.PowerSlots);
```
Comment (Elijahrane) [inline Content.Server/Abilities/Psionics/PsionicAbilitiesSystem.cs:318]: Make 10 a named const. Also this seems way too high

File: Resources/Prototypes/Roles/Jobs/Science/research_director.yml [hunk 2]
```diff
@@ -29,6 +29,7 @@
     components:
     - type: BibleUser # Nyano - Lets them heal with bibles
     - type: Psionic # Nyano - They start with telepathic chat
+      powerSlots: 3
```
Comment (Elijahrane) [inline Resources/Prototypes/Roles/Jobs/Science/research_director.yml:32]: Don't forget his ifrit

**Comment by VMSolidus** [comment]
Alright, this is ready to be reviewed and/or merged.

File: Resources/Prototypes/Roles/Jobs/Science/research_director.yml (hunk 2 above)
Comment (VMSolidus) [reply Resources/Prototypes/Roles/Jobs/Science/research_director.yml:32]: Ifrit can't actually roll powers, so no need. It's defaulted to 1, and the ifrit has 1 power.

**Comment by VMSolidus** [comment]
OPENED TWO WEEKS AGO. THIS GODDAMN PR HAS BEEN UNREVIEWED FOR 2 WEEKS. FUCKING LET ME CONTINUE WORKING ON PSIONICS CODE. REVIEW THIS GODDAMN SHIT.

**Comment by VMSolidus** [comment]
image

PEOPLE HAVE BEEN FUCKING COMPLAINING. I HAVEN'T BEEN ABLE TO CONTINUE WORKING ON PSIONICS IN 2 WEEKS. LET ME CONTINUE WORKING ON MY FUCKING CODE.

**Comment by ShatteredSwords** [comment]
> image
> 
> PEOPLE HAVE BEEN FUCKING COMPLAINING. I HAVEN'T BEEN ABLE TO CONTINUE WORKING ON PSIONICS IN 2 WEEKS. LET ME CONTINUE WORKING ON MY FUCKING CODE.

I read this in AM's voice

**Review by OldDanceJacket — APPROVED** [review]

**Review by Peptide90 — APPROVED** [review]

**Comment by FoxxoTrystan** [comment]
Man im unable to ccorrectly view this PR atm, soon :tm:

Also 2 approved commit merge

**Review by FoxxoTrystan — APPROVED** [review]
I like powerslots.

With all of that its giving me ideas to make the Wizard Antag.

I mean Psionic is the new bluespace, lets just merge it and it The Psionic WIZARD!

I now commit approve.

**Pull request merged**